### PR TITLE
Handle unterminated unicode escapes in regexps

### PR DESCRIPTION
**Pull request by jeremyevans**
This fixes an infinite loop possible after ec3542229b29ec93062e9d90e877ea29d3c19472. For \u{} escapes in regexps, skip validation in the parser, and rely on the regexp code to handle validation. This is necessary so that invalid unicode escapes in comments in extended regexps are allowed.

Fixes [Bug #19750]

### CODE REVIEW

File: test/ruby/test_parse.rb [hunk 1]
```diff
@@ -1052,6 +1052,22 @@ def test_yyerror_at_eol
     assert_syntax_error("    0b\n", /\^/)
   end
 
+  def test_unclosed_unicode_escape_at_eol_bug_19750
+    assert_separately([], "#{<<-"begin;"}\n#{<<~"end;"}")
```
Comment (nobu) [inline test/ruby/test_parse.rb:1056]:
```suggestion
    assert_separately([], "#{<<-"begin;"}\n#{<<~'end;'}")
```

and you can remove doubled backslashes.

File: test/ruby/test_parse.rb [hunk 2]
```diff
@@ -1052,6 +1052,22 @@ def test_yyerror_at_eol
...
+    begin;
+      assert_syntax_error("\\/\\\\u", /too short escape sequence/)
+      assert_syntax_error("\\/\\\\u{", /unterminated regexp meets end of file/)
+      assert_syntax_error("\\/\\\\u{\\\\n", /invalid Unicode list/)
+      assert_syntax_error("\\/a#\\\\u{\\\\n/", /invalid Unicode list/)
+      re = eval("\\/a#\\\\u{\\n$/x")
+      assert_match(re, 'a')
+      assert_not_match(re, 'a#')
+      re = eval("\\/a#\\\\u\\n$/x")
```
Comment (nobu) [inline test/ruby/test_parse.rb:1065]: Also, are `/`s needed to escape?
```suggestion
      assert_syntax_error("/\\u", /too short escape sequence/)
      assert_syntax_error("/\\u{", /unterminated regexp meets end of file/)
      assert_syntax_error("/\\u{\\n", /invalid Unicode list/)
      assert_syntax_error("/a#\\u{\\n/", /invalid Unicode list/)
      re = eval("/a#\\u{\n$/x")
      assert_match(re, 'a')
      assert_not_match(re, 'a#')
      re = eval("/a#\\u\n$/x")
```

File: parse.y [hunk 3]
```diff
@@ -7296,44 +7296,71 @@ tokadd_utf8(struct parser_params *p, rb_encoding **encp,
...
+            for (c = *++p->lex.pcur; p->lex.pcur < p->lex.pend; c = *++p->lex.pcur) {
+                if (c == close_brace) {
+                    tokadd(p, c);
+                    ++p->lex.pcur;
+                    break;
+                }
+                else if (c == term) {
+                    break;
+                }
+                if (c == '\\' && p->lex.pcur + 1 < p->lex.pend) {
```
Comment (nobu) [inline parse.y:7317]: Byte-wise scan for `\` can cause unexpected behavior with some encoding, e.g. Shift_JIS.

File: parse.y [hunk 4]
```diff
@@ -7296,44 +7296,71 @@ tokadd_utf8(struct parser_params *p, rb_encoding **encp,
...
+                    tokadd(p, c);
+                    c = *++p->lex.pcur;
+                }
+                tokadd(p, c);
```
Comment (nobu) [inline parse.y:7321]:
```suggestion
                tokaddmbc(p, c);
```

File: parse.y (hunk 3 above)
Comment (jeremyevans) [reply parse.y:7317]: Can you let me know if there is a better solution?

Comment (nobu) [reply parse.y:7317]: `tokadd_mbchar` is it, but forgot to mention the declaration.

Comment (jeremyevans) [reply parse.y:7317]: No problem, thanks for fixing this.  I'll push a commit for the declaration.

**Review by nobu — APPROVED** [review]

**Pull request merged**